segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameBuilder.java

```diff
@@ -9,36 +9,43 @@
  */
 package io.pravega.segmentstore.server.logs;
 
+import com.google.common.base.Preconditions;
 import io.pravega.common.ExceptionHelpers;
 import io.pravega.common.Exceptions;
-import io.pravega.common.function.CallbackHelpers;
-import io.pravega.common.function.ConsumerWithException;
+import io.pravega.common.ObjectClosedException;
+import io.pravega.common.util.SortedIndex;
 import io.pravega.segmentstore.server.LogItem;
 import io.pravega.segmentstore.storage.DurableDataLog;
 import io.pravega.segmentstore.storage.LogAddress;
-import com.google.common.base.Preconditions;
-import lombok.extern.slf4j.Slf4j;
-
 import java.io.IOException;
 import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
 import java.util.function.Consumer;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
 
 /**
  * Builds DataFrames from LogItems. Splits the serialization of LogItems across multiple Data Frames, if necessary,
  * and publishes the finished Data Frames to the given DataFrameLog.
  */
 @Slf4j
+@NotThreadSafe
 class DataFrameBuilder<T extends LogItem> implements AutoCloseable {
     //region Members
 
-    private static final Duration DATA_FRAME_WRITE_TIMEOUT = Duration.ofSeconds(30); // TODO: actual timeout.
     private final DataFrameOutputStream outputStream;
     private final DurableDataLog targetLog;
-    private final ConsumerWithException<DataFrameCommitArgs, Exception> dataFrameCommitSuccessCallback;
-    private final Consumer<Throwable> dataFrameCommitFailureCallback;
-    private boolean closed;
+    private final Args args;
+    private final AtomicBoolean closed;
     private long lastSerializedSequenceNumber;
     private long lastStartedSequenceNumber;
+    private final AtomicReference<Throwable> failureCause;
```

Comment:
Well, the class itself is not thread safe. `append` and `flush` cannot be called concurrently or bad things will happen. The reason I set this to AtomicReference is because it is most likely set via a callback. We submit the frame to Tier1 and then return synchronously. However we do attach an async exception listener to that call and if it fails, it sets the exception on this object. Since that's likely on a different thread, I wanted to make sure the value is properly set before the failure callback is invoked and someone else might need it.